File: src/ExpoAlarmsModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
  Alarm,
  ExpoAlarmsModuleEvents,
  PermissionsResponse,
  ScheduleAlarmParams,
  UpdateAlarmParams,
} from './ExpoAlarms.types';

declare class ExpoAlarmsModule extends NativeModule<ExpoAlarmsModuleEvents> {
  // Alarm CRUD operations
  scheduleAlarm(params: ScheduleAlarmParams): Promise<Alarm>;
  getAlarmById(id: string): Promise<Alarm | null>;
  getScheduledAlarms(): Promise<Alarm[]>;
  updateAlarm(id: string, params: UpdateAlarmParams): Promise<Alarm>;
  cancelAlarm(id: string): Promise<boolean>;
  cancelAllAlarms(): Promise<void>;

  // Alarm actions
  snoozeAlarm(id: string, duration?: number): Promise<void>;
  dismissAlarm(id: string): Promise<void>;

  // Permissions
  requestPermissions(): Promise<PermissionsResponse>;
  getPermissionsStatus(): Promise<PermissionsResponse>;
}

export default requireNativeModule<ExpoAlarmsModule>('ExpoAlarms');

File: android/src/main/java/expo/modules/alarms/ExpoAlarmsModule.kt
package expo.modules.alarms

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAlarmsModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoAlarms")

        Events("onAlarmTriggered", "onAlarmDismissed", "onAlarmSnoozed")

        // Stubbed functions - Android implementation coming later
        AsyncFunction("scheduleAlarm") { _: Map<String, Any?> -> null }
        AsyncFunction("getAlarmById") { _: String -> null }
        AsyncFunction("getScheduledAlarms") { emptyList<Any>() }
        AsyncFunction("updateAlarm") { _: String, _: Map<String, Any?> -> null }
        AsyncFunction("cancelAlarm") { _: String -> false }
        AsyncFunction("cancelAllAlarms") { }
        AsyncFunction("snoozeAlarm") { _: String, _: Int? -> }
        AsyncFunction("dismissAlarm") { _: String -> }
        AsyncFunction("requestPermissions") { mapOf("status" to "denied") }
        AsyncFunction("getPermissionsStatus") { mapOf("status" to "denied") }
    }
}
